spec/integration/header_parser_spec.rb

```diff
@@ -15,26 +16,102 @@ def to_dto
           end
         end
       end
+
       context 'without headers' do
         let(:headers) { nil }
         it { expect { to_dto }.to raise_error(RequestHandler::MissingArgumentError) }
       end
+
       context 'with headers set correctly' do
         let(:headers) { { 'HTTP_APP_KEY' => 'some.app.key', 'HTTP_USER_ID' => '345' } }
         it { expect(to_dto).to eq(OpenStruct.new(headers: { app_key: 'some.app.key', user_id: '345' })) }
       end
     end
   end
 
+  shared_examples 'it validates the headers' do
+    subject(:to_dto) { class_with_headers_schema.new(request: request).to_dto }
+    let(:request) { build_mock_request(params: {}, headers: headers, body: '') }
+
+    context 'HeaderParser' do
+      let(:class_with_headers_schema) do
+        Class.new(RequestHandler::Base) do
+          options do
+            headers do
+              schema(Dry::Schema.Params do
+                required(:client_id).filled(:string)
+              end)
+            end
+          end
+
+          def to_dto
+            OpenStruct.new(
+              headers: headers
+            )
+          end
+        end
+      end
+
+      context 'when the headers are invalid' do
+        let(:headers) { { 'HTTP_CLIENT_ID' => 123 } }
+
+        it 'raises an exception' do
+          expect { to_dto }.to raise_error(RequestHandler::ExternalArgumentError) do |raised_error|
+            expect(raised_error.errors).to eq(
+              [
+                {
+                  status: '400',
+                  code: 'INVALID_HEADER',
+                  detail: 'must be a string',
+                  source: { header: 'Client-Id' }
+                }
+              ]
+            )
+          end
+        end
+      end
+
+      context 'when the headers are missing' do
+        let(:headers) { {} }
+
+        it 'raises an exception' do
+          expect { to_dto }.to raise_error(RequestHandler::ExternalArgumentError) do |raised_error|
+            expect(raised_error.errors).to eq(
+              [
+                {
+                  status: '400',
+                  code: 'MISSING_HEADER',
+                  detail: 'is missing',
+                  source: { header: 'Client-Id' }
+                }
+              ]
+            )
+          end
+        end
+      end
+
+      context 'when the headers are valid' do
+        let(:client_id) { 'foo.123' }
+        let(:headers) { { 'HTTP_CLIENT_ID' => client_id } }
+
+        it 'does not raise an exception' do
+          expect(to_dto).to eq(OpenStruct.new(headers: { client_id: client_id }))
+        end
+      end
```

Comment:
please also add an example which shows that headers are not only validated but also coerced.
So if the schema applies coercion to some field, to_dto should contain the result of that coercion.